File: SKI.kt
import kotlin.collections.Set

sealed class Structure { abstract override fun toString(): String }
class Apply(val target: Structure, val expr: Structure): Structure() {
    fun s(t: Structure) = if (t is Var || t is SKI) "$t" else "($t)"
    override fun toString() = "${s(target)}${s(expr)}"
}
class Var(val name: String): Structure() { override fun toString() = name }
class Lambda(val arg: String, val expr: Structure): Structure() {
    override fun toString() = "λ$arg.$expr"
}

sealed class SKI(val name: String): Structure() { override fun toString() = name }
object S: SKI("S")
object K: SKI("K")
object I: SKI("I")
object Debug: SKI("D")

fun free(x: String, t: Structure): Boolean {
    fun fv(e: Structure): Set<String> = when(e) {
        is Var -> setOf(e.name)
        is Lambda -> fv(e.expr) subtract setOf(e.arg)
        is Apply -> fv(e.target) union fv(e.expr)
        else -> setOf()
    }
    return fv(t).contains(x)
}

fun Structure.toSKI(): Structure = when (this) {
    is Var, is SKI -> this
    is Apply -> Apply(target.toSKI(), expr.toSKI())
    is Lambda -> {
        if (!free(arg, expr)) Apply(K, expr.toSKI())
        else if (expr is Var && expr.name == arg) I
        else if (expr is Lambda && free(arg, expr.expr)) Lambda(arg, expr.toSKI()).toSKI()
        else if (expr is Apply && (
            free(arg, expr.target) || free(arg, expr.expr)
        )) Apply(
            Apply(S, Lambda(arg, expr.target.toSKI()).toSKI()),
            Lambda(arg, expr.expr.toSKI()).toSKI()
        )
        else Debug
    }
}

fun main() {
    val applier: (String) -> Lambda = { q -> Lambda(q, Apply(Var(q), Var(q))) }
    val t = Lambda("x", Apply(Var("f"), applier("x")))
    val f = Lambda("f", Apply(t, t))
    println(f)
    println(f.toSKI())
}

File: SKI.py
from collections import namedtuple

L = type('L', (namedtuple('_L', 'var expr'),), {
  '__repr__': lambda self: f'λ{self.var}.{self.expr}'
})
Apply = type('Apply', (namedtuple('_Apply', 'target expr'),), {
  's': lambda self, x: str(x) if isinstance(x, (Var, CL)) else f'({x})',
  '__repr__': lambda self: f'{self.s(self.target)}{self.s(self.expr)}'
})
Var = type('Var', (namedtuple('_Var', 'name'),), {
  '__hash__': lambda self: hash(self.name),
  '__repr__': lambda self: self.name
})
CL = namedtuple('CL', 'name')
setattr(CL, '__repr__', lambda self: self.name)
S = CL('S')
K = CL('K')
I = CL('I')

def free(x, e):
  def fv(x):
    if isinstance(x, Var): return {x}
    if isinstance(x, Apply): return fv(x.target) | fv(x.expr)
    if isinstance(x, L): return fv(x.expr) - {Var(x.var)}
    return set()
  return Var(x) in fv(e)

def T(x):
  if isinstance(x, (Var, CL)): return x
  if isinstance(x, Apply): return Apply(T(x.target), T(x.expr))
  if isinstance(x, L):
    v, expr = x.var, x.expr
    if not free(v, expr): return Apply(K, T(expr))
    if expr == Var(v): return I
    if isinstance(expr, L) and free(v, expr.expr):
      return T(L(v, T(expr)))
    if isinstance(expr, Apply) and (free(v, expr.target) or free(v, expr.expr)):
      print(L(v, T(expr.target)), L(v, T(expr.expr)))
      return Apply(
        Apply(S, T(L(v, T(expr.target)))),
        T(L(v, T(expr.expr)))
      )

W = lambda q: Apply(Var(q), Var(q))
F = L('x', Apply(Var('f'), W('x')))
a = L('f', Apply(L('t', W('t')), F))
print(a)
print(T(a))
